src/hotspot/cpu/riscv/gc/x/x_riscv.ad

```diff
@@ -65,17 +65,17 @@ instruct xLoadP(iRegPNoSp dst, memory mem, rFlagsReg cr)
   ins_encode %{
     const Address ref_addr (as_Register($mem$$base), $mem$$disp);
     __ ld($dst$$Register, ref_addr);
-    x_load_barrier(_masm, this, ref_addr, $dst$$Register, t1 /* tmp */, barrier_data());
+    x_load_barrier(_masm, this, ref_addr, $dst$$Register, $tmp$$Register /* tmp */, barrier_data());
   %}
 
   ins_pipe(iload_reg_mem);
 %}
 
-instruct xCompareAndSwapP(iRegINoSp res, indirect mem, iRegP oldval, iRegP newval, rFlagsReg cr) %{
+instruct xCompareAndSwapP(iRegINoSp res, indirect mem, iRegP oldval, iRegP newval, iRegINoSp tmp, rFlagsReg cr) %{
   match(Set res (CompareAndSwapP mem (Binary oldval newval)));
   match(Set res (WeakCompareAndSwapP mem (Binary oldval newval)));
   predicate(UseZGC && !ZGenerational && !needs_acquiring_load_reserved(n) && n->as_LoadStore()->barrier_data() == XLoadBarrierStrong);
-  effect(KILL cr, TEMP_DEF res);
+  effect(KILL cr, TEMP_DEF res, TEMP tmp);
```

Comment:
You might want to remove `KILL cr` from `effect` if `cr` (aka `t1`) is not used anymore after this change.

Reply:
I don't *think* that matters, as masm may use t1 so this 'may' kill cr.
But since cr never contains any actual information and is never read (as cr), it's unclear to me, if we ever can get unwanted side-effect if either missing a "KILL cr" or have one to many, you know?

As we have fused cmp+branch I would think the graph would be better of believing we always have a live CR.
We don't need to re-do the cmp to get back the killed cr, we can pretend it is still alive, since the fused branch will act as it is alive, no?

Also I have patch which cleans-up the registers in riscv.ad, I'll try to get it out as RFC in form of a PR. (i.e. my suggestions)

Reply:
> I don't _think_ that matters, as masm may use t1 so this 'may' kill cr. But since cr never contains any actual information and is never read (as cr), it's unclear to me, if we ever can get unwanted side-effect if either missing a "KILL cr" or have one to many, you know?
> 
> As we have fused cmp+branch I would think the graph would be better of believing we always have a live CR. We don't need to re-do the cmp to get back the killed cr, we can pretend it is still alive, since the fused branch will act as it is alive, no?

The C2 JIT does expecting something in a CR register. Check the nodes in file riscv.ad like `cmpFastLock`, `cmpFastUnlock` and `partialSubtypeCheckVsZero` [1]. That's way we emulated the CR register on riscv using the `t1` general-purpose register. I don't see another way without changing the C2's assumption about CR. 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/riscv.ad#L10434

Reply:
Sorry, I'm not saying we should remove cr, I'm just saying I don't think need to KILL cr.
C2 will think cr is valid all the time.

Reply:
I tested removing all "KILL cr" from all risc.ad files. Compiler tests c2/codegen to works fine (not 100% done yet).

Which was my theory, as we never consume or produce values in 'cr' we can see cr as always having correct values, which is nothing.

Reply:
This is not related to this bug (yes, we can remove KILL cr):

Ah, this is very problematic, as t1 is used masm, which means calling masm may destroy cr.

For exampel C2 may call:
`MacroAssembler::zero_words(Register ptr, Register cnt)` as it do not clobber t1 but if we would call:
`MacroAssembler::zero_words(Register base, uint64_t cnt) ` which do clobber t1 suddenly we need to add KILL cr.

Which means if someone just do a small change like:
- masm->zero_words(reg1, reg2);
+ masm->zero_words(reg1, 64);

We suddenly can get very subtle bugs.

Reply:
Yeah, I agree it is error-prone. But it is the same case for ther other platforms which have a real CR register, isn't it? They should have the same issue. I guess it doesn't deserve it if we keep `t1` the dedicated CR register, that is not touched else where by other masm assembers.

Reply:
I think we can pretty easy add a scoped object, similar to uncompressed region, where we can black-list registers.

So the ad file would add for example to fast_lock:

```
{
  AllowAdditionalRegister(t1); // As we kill cr fastlock may use t1/(cr) in addition to t0 and passed in regs.
  ___ fastlock(...);
}
```

Reply:
Anyhow something to think about.

Reply:
Your proposal sounds interesting. It's like a way of self-checking/assertion about `t1` register usage. The masm could stop when it sees usage of `t1` in a scope without an `AllowAdditionalRegister(t1)`. But I am not sure if there is a good place for us to add checking for that purpose.

Reply:
@robehn @RealFYang I have done a check and no cr is being used and have removed the extra kill cr. 
please take another look.